Refactor product-control to TypeScript, clean up related types.

samueljseay opened the pull request:
## What

Convert product-control to TypeScript, remove prop types and clean up related types.

Fixes #9537

## Why

This is desired to refactor existing JS components to TS.

I've opted not to add additional tests here, there is test coverage of both the HOCs it uses and other underlying components and we can also perform some manual testing of the block.

## Testing Instructions

1. A smoke test can involve creating a page and testing the functionality of the single product block, e.g. search for product, choose product and check that it still works. 

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [x] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Add suggested changelog entry here.

samueljseay committed:
Refactor product-control to TypeScript, clean up related types.

samueljseay committed:
Tell eslint we are in browser env to pick up on Response object in tests.

samueljseay committed:
Improve the types of formatError.

samueljseay committed:
Use Response object to improve formatError tests and remove type errors.

samueljseay committed:
Fix up type errors where possible, ignore some that need refactoring.

samueljseay commented on .eslintrc.js:
```diff
@@ -117,6 +117,9 @@ const restrictedImports = [
 ];
 
 module.exports = {
+	env: {
```
Allows us to use Response global in Jest test

samueljseay commented on assets/js/base/utils/errors.ts:
```diff
@@ -0,0 +1,41 @@
+export interface ErrorObject {
```
Move the ErrorObject here, it makes sense to be here as a place for reuse imo.

samueljseay commented on assets/js/editor-components/product-control/index.tsx:
```diff
@@ -39,21 +44,39 @@ const messages = {
 	error,
 	instanceId,
-	isCompact,
+	isCompact = false,
```
default args to match the proptypes defaults.

samueljseay commented on assets/js/editor-components/product-control/index.tsx:
```diff
@@ -142,29 +165,37 @@ const ProductControl = ( {
 		} else if ( showVariations ) {
 			return renderItemWithVariations;
 		}
-		return null;
```
This is a behaviour change, but returning null was wrong, we attempt to call this as a function in underlying components. I guess its possible it never came up if the first 2 conditions are always met.

samueljseay commented on assets/js/editor-components/product-control/index.tsx:
```diff
@@ -142,29 +165,37 @@ const ProductControl = ( {
 
 	const currentVariations =
-		variations && variations[ expandedProduct ]
+		variations && expandedProduct !== null && variations[ expandedProduct ]
```
This won't change behaviour, since indexing an object by null will return undefined.. unless it had a prop called `null`. 😰

samueljseay committed:
Support string includes as well in if check.

samueljseay commented on assets/js/editor-components/product-control/index.tsx:
```diff
@@ -142,29 +165,42 @@ const ProductControl = ( {
-				selected.includes( id )
-			) }
+			// @ts-expect-error - It looks like ProductResponse does not fulfill the SearchListItemProps interface. This needs to be fixed.
+			selected={ currentList.filter( ( { id } ) => {
```
Based on types it's really hard to know what to do here. selected can be string | number[] | number according to the  type.

if its `number` then includes is not callable, if its array or string then it is and thats ok.

samueljseay commented on assets/js/editor-components/search-list-control/item.tsx:
```diff
@@ -22,13 +22,22 @@ const Count = ( { label }: { label: string | React.ReactNode | number } ) => {
+	breadcrumbs: string[];
+};
+
+const checkHasBreadCrumbs = (
```
Typeguard allows us to properly narrow the type to call `getBreadcrumbsForDisplay`

samueljseay commented on assets/js/editor-components/search-list-control/types.ts:
```diff
@@ -23,7 +23,7 @@ interface SearchListProps {
 	list: SearchListItem[];
 	// Callback to render each item in the selection list, allows any custom object-type rendering.
-	renderItem?: ( args: renderItemArgs ) => JSX.Element;
+	renderItem?: ( args: renderItemArgs ) => JSX.Element | null;
```
Looking at usages I think its valid this can return null

samueljseay committed:
Maybe count and value are required.

samueljseay committed:
Fix up ts ignore comment.

samueljseay commented on assets/js/base/utils/errors.ts:
```diff
@@ -0,0 +1,41 @@
+export const formatError = async (
+	error: SimpleError | Response
+): Promise< ErrorObject > => {
+	if ( 'json' in error ) {
```
This error formatter is covered by tests, so I feel confident to make some small runtime changes here.

samueljseay committed:
Remove redundant ts ignore.

samueljseay commented on assets/js/editor-components/search-list-control/types.ts:
```diff
@@ -35,7 +35,7 @@ export interface ListItemsProps
 
 export type SearchListItem = {
-	breadcrumbs: string[];
+	breadcrumbs?: string[];
```
It's a guess but, based on original checks [here](https://github.com/woocommerce/woocommerce-blocks/pull/10698/files#diff-40b9b2d5974a47d8303bf833d28d3820b0f8911fadf8ab84a2552e569fea97a2L27), this may not be present.

dinhtungdu replied:
According to [this](https://github.com/woocommerce/woocommerce-blocks/blob/0ade96139c899d39d9bc56386a5eba7fbe88a008/assets/js/utils/attributes.ts#L63), this property will always exist.

samueljseay committed:
Merge branch 'trunk' into dev/refactor-prod-control-ts

samueljseay committed:
Use alias to import ErrorObject type.

samueljseay committed:
Ignore type issue in usage of withProductVariations, fix import.

samueljseay committed:
Remove js suffix from imports of error.ts

samueljseay committed:
Merge branch 'trunk' into dev/refactor-prod-control-ts

samueljseay commented on assets/js/base/utils/test/errors.js:
```diff
@@ -17,9 +20,9 @@ describe( 'formatError', () => {
-		const error = await formatError( {
-			json: () => Promise.resolve( { message: 'Lorem Ipsum' } ),
-		} );
+		const mockResponse = new Response( mockResponseBody, { status: 400 } );
```
Use Response to pass more realistic argument to formatError. a plain object that has property of `json` is not type safe or close to real usage.

samueljseay commented on assets/js/editor-components/search-list-control/types.ts:
```diff
@@ -114,7 +114,7 @@ export interface SearchListControlProps {
 	onSearch?: ( ( search: string ) => void ) | undefined;
 	// Callback to render each item in the selection list, allows any custom object-type rendering.
-	renderItem?: ( args: renderItemArgs ) => JSX.Element;
+	renderItem?: ( args: renderItemArgs ) => JSX.Element | null;
```
Based on usages it seems valid this could also return null in some cases. null is valid to return when rendering JSX

dinhtungdu commented on assets/js/editor-components/product-attribute-term-control/index.tsx:
```diff
@@ -65,7 +65,7 @@ const ProductAttributeTermControl = ( {
 		];
 
-		if ( ! item.breadcrumbs.length ) {
+		if ( ! item?.breadcrumbs?.length ) {
```
If `breadcrumbs` always exists, then we need to update this and other optional chaining calls.

dinhtungdu reviewed (commented):
Look really good to me 😄. I just have a question around the type of breadcrumb (and the associated optional chaining). Otherwise, this is ready to be merged.

samueljseay committed:
Breadcrumbs should always at least be empty array.

samueljseay committed:
Merge branch 'trunk' into dev/refactor-prod-control-ts

samueljseay commented:
@dinhtungdu Thanks for the feedback. I'm a little nervous about the type since removing optional chaining will change runtime behaviour and I'm concerned we don't have enough tests to cover the cases, that said, I think I agree it **should** always be present so I've changed to mandatory, removed optional chaining and typecheck here: https://github.com/woocommerce/woocommerce-blocks/pull/10698/commits/adda87fcca17f678f9426c56cb9fa0fee12d4db5

Interestingly, none of the unit tests for the SearchListControl pass breadcrumbs, but not sure if thats actual usage or just oversight/omission  in testing

samueljseay commented:
@dinhtungdu Interestingly, none of the unit tests for the SearchListControl pass breadcrumbs, but not sure if thats actual usage or just oversight/omission  in testing. I can update them, but are you really sure about this? i am concerned the optional chaining is there for a reason. I think perhaps its not wise to change runtime behaviour, but rather put a TODO on the type to find out if indeed breadcrumbs are always present?

A type says one thing, but runtime code says another so I don't think the code you linked necessarily verifies what the type is?

dinhtungdu commented:
@samueljseay 

> I can update them, but are you really sure about this?

I am not.

> I think perhaps its not wise to change runtime behaviour, but rather put a TODO on the type to find out if indeed breadcrumbs are always present?

I agree. Changing runtime behavior is out of scope for a refactoring PR as well. 

> A type says one thing, but runtime code says another so I don't think the code you linked necessarily verifies what the type is?

I think I was wrong here because I can find another [list](https://github.com/woocommerce/woocommerce-blocks/blob/7e8e622227d8a3c031e2cd117d9aecfb9586f934/assets/js/blocks/attribute-filter/edit.tsx#L160-L167) that doesn't have breadcrumb. I'm sorry for the false alarm.

opr committed:
Move cart widget tests to playwright and ensure classic/block theme configs are executed correctly (#10669)

kmanijak committed:
Add Product Collection Full Grid pattern (#10665)

* Add initial look of Full Grid Product Collection pattern

* Refactor textAlign property of Product Button so it uses flex rather than text-align

* Update Product Button text align after the fix

* Remove debug log (#10719)

* Add Product Collection Grid pattern (#10660)

* Dispatch the `wc-blocks_render_blocks_frontend` event when rendering the empty cart block (#10619)

* E2E: Refactor Mini Cart to be ready for fully parallel (#10704)

* Revert dequeue add-to-cart-variation script which is needed to properly handle variable products in single product page (#10723)

* Update alignment options to new setting

---------

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Tung Du <dinhtungdu@gmail.com>

wavvves committed:
Display notices on cart/checkout FSE templates (#10151)

* Display notices on cart/checkout FSE templates

* revert space removal

Commit:
Update integration-interface.md

Minor changes to wording, where "both" was used to refer to 3 things

gigitux committed:
Product Gallery Block: Next/Previous block (#10235)

* Add Product Gallery Next Previous block

* update description

* update registration

* improve style

* use context instead of attributes

* add eslint and ts lint exception

* improve CSS

* E2E Next Previous block (#10329)

* Add E2E tests

* fix e2e tests path after rebase

* add screenshots

* improve E2E test

* improve E2E test

* improve E2E test

---------

Co-authored-by: Manish Menaria <the.manish.menaria@gmail.com>

* fix webpack conf

* Next Previous Button block - Add layout support (#10668)

* Add layout support

* Product Gallery Next Previous: Fix block crashing in the editor when vertical alignment has not been customized yet

* Product Gallery Next Previous: Position fixes for the Next/Previous button both in the editor and frontend

* add E2E test

* fix import

---------

Co-authored-by: Daniel Dudzic <daniel.dudzic@automattic.com>

* Next Previous Buttons: hide buttons when the product has only one image (#10743)

* Next Previous Buttons: hide buttons when the product has only one image

* update URL

* fix import

* set default value

* fix E2E test

---------

Co-authored-by: Manish Menaria <the.manish.menaria@gmail.com>
Co-authored-by: Daniel Dudzic <daniel.dudzic@automattic.com>

dinhtungdu committed:
E2E: ensure the shop page is created by creating it manually (#10744)

sunyatasattva committed:
Fix inconsistent padding in Customer Account block (#10476)

sunyatasattva committed:
Convert `product-control` to TypeScript (#10552)

The main aim of this PR was to convert the product-control component to TypeScript. However, since the component is coupled with many other parts (it is itself wrapped into 4 HOCs), many refactoring changes were needed to make the types work.

In particular:

* Add `convertProductResponseItemToSearchItem` utility function

The function allows to clean up the `ProductResponseItem` in order
to make its shape compatible with the `SearchListItem`, instead of
passing the entire object.

Because data from the object might need to be used in certain
circumstance (e.g. render functions), the data is now allowed to be
scoped within a `details` key.

This commit also converts `utils/products` to TypeScript.

* Fix `ProductResponseItem` type missing `slug`

The `slug` is returned by the API but didn't appear in the type def.

* Improve certain HOCs types

The `withSearchedProducts` and `withTransformSingleSelectToMultipleSelect`
HOCs are now typed in such a way that it is clear what incoming props
they expect and what they inject.

* Refactor `ProductAttrributeTermControl` for further type safety

Make sure that `count` is always a `number` as required by the
component types.

* Refactor `search-list-control.tsx` and `item.tsx`

The new `SearchItem` allows for the `details` key to be passed.
All the components are now aware of it.

samueljseay commented:
Looks like I missed that this was already assigned to someone else and completed in https://github.com/woocommerce/woocommerce-blocks/pull/10552

Whoops!
